csharp/src/Ice/Identity.cs

```diff
@@ -0,0 +1,134 @@
+            // backslash before the separator.
+            //
+            int slash = -1, pos = 0;
+            while ((pos = s.IndexOf('/', pos)) != -1)
+            {
+                int escapes = 0;
+                while (pos - escapes > 0 && s[pos - escapes - 1] == '\\')
+                {
+                    escapes++;
+                }
+
+                //
+                // We ignore escaped escapes
+                //
+                if (escapes % 2 == 0)
+                {
+                    if (slash == -1)
+                    {
+                        slash = pos;
+                    }
+                    else
+                    {
+                        //
+                        // Extra unescaped slash found.
+                        //
+                        throw new ArgumentException($"unescaped backslash in identity `{s}'");
+                    }
+                }
+                pos++;
+            }
+
+            string category;
+            string? name = null;
+            if (slash == -1)
+            {
+                try
+                {
+                    name = IceUtilInternal.StringUtil.unescapeString(s, 0, s.Length, "/");
+                }
+                catch (ArgumentException ex)
+                {
+                    throw new FormatException($"invalid name in identity `{s}", ex);
+                }
+                category = "";
+            }
+            else
+            {
+                try
+                {
+                    category = IceUtilInternal.StringUtil.unescapeString(s, 0, slash, "/");
+                }
+                catch (ArgumentException ex)
+                {
+                    throw new FormatException($"invalid category in identity `{s}", ex);
+                }
+
+                if (slash + 1 < s.Length)
+                {
+                    try
+                    {
+                        name = IceUtilInternal.StringUtil.unescapeString(s, slash + 1, s.Length, "/");
+                    }
+                    catch (ArgumentException ex)
+                    {
+                        throw new FormatException($"invalid name in identity `{s}", ex);
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new FormatException($"empty name in identity `{s}'");
+            }
+
+            return new Identity(name, category);
+        }
+
+        public static bool TryParse(string s, out Identity? identity)
+        {
+            try
+            {
+                identity = Parse(s);
+                return true;
+            }
+            catch (Exception)
+            {
+                identity = null;
+                return false;
+            }
+        }
+
+        public override string ToString()
+        {
+            return ToString(ToStringMode.Unicode);
+        }
+
+        public string ToString(Communicator communicator)
```

Comment:
A ToString overload that takes a ToStringMode would be nicer.

Reply:
It is below that, should I remove this one?

Reply:
It is below that, should I remove this one?

Reply:
I missed it. Yes, let's remove this communicator overload.